File: 03-core-periphery/src/Store.ts
// 03-core-periphery/src/Store.ts
// The corrected user store. The UserRecord type, the seed users, and the hash
// helper are given in ./store-data and are not part of what you edit here.

import { SEED_USERS, UserRecord, hash } from "./store-data";

export interface PublicProfile {
  username: string;
  role: string;
}

// The narrow interface: the only way anything outside the core reaches it.
// Every method is hash-free, so a holder of this interface has no path to a hash.
export interface UserDirectory {
  verify(username: string, password: string): boolean;
  getPublicProfile(username: string): PublicProfile | undefined;
  usernames(): string[];
}

// Core: the only place the hash lives. `users` is private and the class exposes
// nothing beyond UserDirectory, so the hash cannot be reached from outside.
export class UserStore implements UserDirectory {
  private users: Record<string, UserRecord> = SEED_USERS;

  verify(username: string, password: string): boolean {
    const user = this.users[username];
    if (!user) return false;
    return user.passwordHash === hash(password);
  }

  getPublicProfile(username: string): PublicProfile | undefined {
    const user = this.users[username];
    if (!user) return undefined;
    return { username: user.username, role: user.role };
  }

  usernames(): string[] {
    return Object.keys(this.users);
  }

  // Admin convenience: hand back a user's full record. Guarded by a role check.
  getUser(requesterRole: string, username: string): UserRecord | undefined {
    if (requesterRole !== "admin") return undefined;
    return this.users[username];
  }
}

// Periphery: a free function holding only the narrow interface. It has no
// reference to `users` and no method that yields a hash, so there is no code
// path from here to the secret.
export function report(directory: UserDirectory): string[] {
  return directory.usernames().map((name) => {
    const profile = directory.getPublicProfile(name)!;
    return profile.username + " (" + profile.role + ")";
  });
}

File: 03-core-periphery/src/Store.cs
// 03-core-periphery/src/Store.cs
// A minimal user store. Plain C#, no framework.
using System.Collections.Generic;

public record UserRecord(string Username, string PasswordHash, string Role);

public class UserStore
{
    private readonly Dictionary<string, UserRecord> _users = new()
    {
        ["alice"] = new UserRecord("alice", "hash:alice-pw", "admin"),
        ["bob"] = new UserRecord("bob", "hash:bob-pw", "member"),
    };

    public bool Verify(string username, string password)
    {
        if (!_users.TryGetValue(username, out var user)) return false;
        return user.PasswordHash == Hash(password);
    }

    public UserRecord? GetUser(string username)
    {
        return _users.TryGetValue(username, out var user) ? user : null;
    }

    public List<string> Report()
    {
        var rows = new List<string>();
        foreach (var name in _users.Keys)
        {
            var user = GetUser(name)!;
            rows.Add($"{user.Username} ({user.Role}) {user.PasswordHash}");
        }
        return rows;
    }

    private static string Hash(string password) => $"hash:{password}";
}
